### PR TITLE
consolidate version information

### PR DESCRIPTION
**Pull request by radditude**
Now that we've switched to our new build system and are using [`version/VERSION`](https://github.com/hashicorp/terraform/blob/main/version/VERSION) as the canonical way to discover versions during the build process, it is unfortunately rather easy for `version/VERSION` and the version defined in [`version/version.go`](https://github.com/hashicorp/terraform/blob/main/version/version.go) to drift out of sync, which causes unfortunate problems like #32963 at release time. 

This PR proposes making `version/VERSION` the canonical source for version info at all times, using a [file embed](https://github.com/hashicorp/terraform/blob/main/version/version.go). To make things less confusing in local development, we'll keep the `-dev` prerelease by default and set the expectation that release builds, whether ours or a third-party build process, must specifically mark themselves as such via a linker flag in order to report the correct version.

## Target Release

1.6.x

### CODE REVIEW

File: version/version.go [hunk 1]
```diff
@@ -5,25 +5,36 @@
...
+// rawVersion is the current version as a string, as read from the VERSION
+// file. This must be a valid semantic version.
+//
+//go:embed VERSION
+var rawVersion string
+
 func init() {
+	rawVersion = strings.TrimSpace(rawVersion)
+	version.Must(version.NewVersion(rawVersion))
+	Version, Prerelease = extractPrerelease(rawVersion)
```
Comment (liamcervante) [inline version/version.go:37]: I like this! But I think you don't need to parse the prerelease yourself. Can you do this?

```
parsed := version.Must(version.NewVersion(rawVersion))
Version = parsed.Core().String()
Prerelease = parsed.Prerelease()
```

File: version/version.go [hunk 2]
```diff
@@ -5,25 +5,36 @@
...
 
 // The main version number that is being run at the moment.
-var Version = "1.5.0"
+var Version string
 
 // A pre-release marker for the version. If this is "" (empty string)
 // then it means that it is a final release. Otherwise, this is a pre-release
 // such as "dev" (in development), "beta", "rc1", etc.
-var Prerelease = "dev"
+var Prerelease string
```
[the remark this answers is not in the document]
Comment (radditude) [reply version/version.go]: Reworked this a bit! Now we'll keep `-dev` for local builds, and for release builds we'll set a linker flag that says, essentially, "report exactly the version in `version/VERSION`".

File: version/version.go (hunk 1 above)
Comment (radditude) [reply version/version.go]: Good point! Reworked this a bit.

**Comment by apparentlymart** [comment]
FWIW the legacy release process includes some non-robust regular expressions looking for the `Version` and `Prerelease` variables in `version.go` and so it might no longer behave correctly after this change is merged.

 I don't think it particularly matters since we have no intention to return to using the legacy process, but if we _do_ end up needing to use the legacy process in any branch where this has been merged we should be sure to check that what it built has a correct version number compiled into it. (In practice I think the e2etests would catch it, because running `terraform version` to see if it returns the expected version number is one of the tests in that suite.)

**Comment by radditude** [comment]
Excellent callout! I didn't preemptively open a PR over in the private repo since we haven't yet agreed that the path proposed in this PR is a good one, but I will certainly do that as followup.

File: version/version.go [hunk 3]
```diff
@@ -5,26 +5,45 @@
...
+// without any prerelease information.
 var SemVer *version.Version
 
 func init() {
-	SemVer = version.Must(version.NewVersion(Version))
+	semVerFull := version.Must(version.NewVersion(strings.TrimSpace(rawVersion)))
+	SemVer = semVerFull.Core()
+	Version = SemVer.String()
+
+	if dev == "no" {
```
Comment (sarahethompson) [inline version/version.go:42]: If there is already a prelease set in VERSION (e.g `-alpha`) would the desired behavior for local builds be that this is dropped in favour of `-dev` or the other way around?

Comment (radditude) [reply version/version.go:42]: Dropping any existing prerelease in favor of `-dev`! That's been our convention previously, so seemed just as well to stick with it for now.

**Comment by radditude** [comment]
I triggered a build from this branch and verified that the binaries produced by it report the correct version number: https://github.com/hashicorp/terraform/actions/runs/5581974639#artifacts

**Pull request merged**